### PR TITLE
Added support for log_model='best_and_last' option in wandb logger

### PR DESCRIPTION
**Pull request by ohayonguy**
## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun? yup :)

Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/loggers/test_wandb.py [hunk 1]
```diff
@@ -176,6 +176,16 @@ def test_wandb_log_model(wandb, tmpdir):
...
 
+    # test log_model='best_and_last'
+    wandb.init().log_artifact.reset_mock()
+    wandb.init.reset_mock()
+    logger = WandbLogger(log_model="best_and_last")
+    logger.experiment.id = "1"
+    logger.experiment.project_name.return_value = "project"
+    trainer = Trainer(default_root_dir=tmpdir, logger=logger, max_epochs=2, limit_train_batches=3, limit_val_batches=3)
+    trainer.fit(model)
+    assert wandb.init().log_artifact.call_count == 2
```
Comment (justusschock) [inline tests/loggers/test_wandb.py:187]: this does not test, that the other ones are properly removed. Can we also test/mock this somehow?

**Comment by justusschock** [comment]
@borisdayma mind reviewing this PR?

File: tests/loggers/test_wandb.py (hunk 1 above)
Comment (ohayonguy) [reply tests/loggers/test_wandb.py:187]: For sure. I will figure out how this should be properly tested.

File: pytorch_lightning/loggers/wandb.py [hunk 2]
```diff
@@ -301,3 +307,13 @@ def _scan_and_log_checkpoints(self, checkpoint_callback: "ReferenceType[ModelChe
...
+
+        if self._log_model == "best_and_last":
+            # Clean up previous artifacts.
+            # Adapted from https://gitbook-docs.wandb.ai/guides/artifacts/api#cleaning-up-unused-versions
+            api = wandb.Api(overrides={"project": self.experiment.project})
+
+            for version in api.artifact_versions(f"model-{self.experiment.id}", "model"):
+                # Clean up all versions that don't have an alias such as 'latest'.
+                if len(version.aliases) == 0:
+                    version.delete()
```
Comment (tchaton) [inline pytorch_lightning/loggers/wandb.py:319]: noob question. I am not familiar with wandb API. 

It seems model are being versioned and your are deleting all versions which doesn't have either latest or best aliases.

I am not sure to grasp why this would save only best and last model weights.

Furthermore, I don't think this would work for multiple ModelCheckpoint. Should we save the monitor as metadata to perform the filtering.

best_and_last should produce at maximum num_model_checkpoints + 1 checkpoints right ?

Comment (borisdayma) [reply pytorch_lightning/loggers/wandb.py:319]: We automatically tag some artifact versions (model checkpoints). We tag the "latest" and we tag the "best" when monitoring value is defined (they can point to the same model). So there are 2 versions tagged at most.

File: tests/loggers/test_wandb.py (hunk 1 above)
Comment (borisdayma) [reply tests/loggers/test_wandb.py:187]: Maybe you should train for 5 epochs in this test

**Review by borisdayma — COMMENTED** [review]
Looks good to me. I'm just curious of the naming and if we could offer this functionality in a cleaner way.

An alternative could be to just add an argument such as "remove_unused_artifacts" to the logger which would be false be default.
Then we can just have a separate function for this purpose.

File: pytorch_lightning/loggers/wandb.py (hunk 2 above)
Comment (ohayonguy) [reply pytorch_lightning/loggers/wandb.py:319]: @borisdayma Yah exactly. The implementation relies on the fact that there are at most 2 live aliases at the same time.
Several aliases can probably be included (best_0, best_1, best_2, ..., latest), but this is another PR :)

**Comment by ohayonguy** [comment]
> Looks good to me. I'm just curious of the naming and if we could offer this functionality in a cleaner way.
> 
> An alternative could be to just add an argument such as "remove_unused_artifacts" to the logger which would be false by default.
> Then we can just have a separate function for this purpose. What do you think?

Great idea @borisdayma to make the naming cleaner. However, I think that we need to take this a step further. So we would pass `log_model=all` and `remove_unused_artifacts=True` to achieve our functionality? So then, what does "all" means, and what does it refer to? It's a bit confusing to me, and I think we need to come up with a solution that is easier to interpret in terms of what's happening in the background. 

Let's examine the current functionality of `log_model`:

- Setting this flag to `True` means to save all the checkpoints ONLY WHEN TRAINING FINISHES. Why is this behavior necessary at all? In what situation would this be preferred over saving checkpoints along the way and deleting old versions, like `ModelCheckpoint` does? The final result is anyways the same, and the disk space usage is also the same (in the end).
- Setting this flag to `all` means to save the entire history of checkpoints.

Therefore, I suggest:

1. Change the naming of `log_model` to `save_checkpoints_as_artifacts`
2. `save_checkpoints_as_artifacts=True` would mean to save checkpoints exactly like ModelCheckpoint. This means that we should add aliases like `best_0`, ... `best_k` in case save_top_k>1 is passed... (what else should we take into consideration here?)
3. Add a flag `keep_artifacts_history`. If this flag is set, then checkpoints are not deleted from the artifacts history.

What do you think?
 Thanks!

**Comment by borisdayma** [comment]
So my concern was related to people with limited network bandwidth, whether upload speed or max quota before being throttled.

I have no issue with changing arguments name but then it requires backward compatibility so better to avoid it unless it's clearly beneficial.

File: pytorch_lightning/loggers/wandb.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/loggers/wandb.py:319]: I believe best_{monitor_0}, best_{monitor_1}, best_{monitor_2} would be better, it would enable users to navigate their weights better on the Wandb UI.

File: tests/loggers/test_wandb.py [hunk 3]
```diff
@@ -176,6 +176,16 @@ def test_wandb_log_model(wandb, tmpdir):
     trainer.fit(model)
     assert wandb.init().log_artifact.call_count == 2
 
+    # test log_model='best_and_last'
+    wandb.init().log_artifact.reset_mock()
+    wandb.init.reset_mock()
+    logger = WandbLogger(log_model="best_and_last")
+    logger.experiment.id = "1"
+    logger.experiment.project_name.return_value = "project"
+    trainer = Trainer(default_root_dir=tmpdir, logger=logger, max_epochs=2, limit_train_batches=3, limit_val_batches=3)
```
Comment (tchaton) [inline tests/loggers/test_wandb.py:185]: Should this test use a ModelCheckpoint ?

File: pytorch_lightning/loggers/wandb.py (hunk 2 above)
Comment (borisdayma) [reply pytorch_lightning/loggers/wandb.py:319]: The idea is that we directly leverage `ModelCheckpoint` to identify best metrics (easier to maintain the callback, avoid replicating the same logic, and maybe easier for users).
You can see an example [here](http://wandb.me/lightning).

**Comment by borisdayma** [comment]
Ok so I really like the purpose of this PR.

I think there are a few things to decide:

* should we be concerned about bandwidth, max upload quota for users? I personally don't have that issue but the reason we have a mode `log_model=True` that uploads only at the end is for that exact purpose
* depending on the answer to previous function, we can rethink allowed values for `log_model` (right now `True` and `all`)
* should the "artifact removal" be on by default? I don't have a strong opinion about it.
* how to integrate it best? I like `keep_artifacts_history` (suggested by @ohayonguy ) as an arg in the callback and a separate function of the same name that will be called when the flag is on.
* I suggest to keep the functionality `best_1`, `best_2`… when wanting to keep top k models in a separate PR as it is more complex and opens many more questions

File: tests/loggers/test_wandb.py (hunk 3 above)
Comment (ohayonguy) [reply tests/loggers/test_wandb.py:185]: Yup. We should definitely build a better test. I am just not familiar with the whole "Mocking" thing, and with tests in general. Should my test function be also decorated with mocking?

If not, then this requires some wandb default setup to run the experiment. Not sure how to make this right as I never wrote tests for such a large project like Lightning :)

**Comment by ohayonguy** [comment]
> Ok so I really like the purpose of this PR.
> 
> I think there are a few things to decide:
> 
> * should we be concerned about bandwidth, max upload quota for users? I personally don't have that issue but the reason we have a mode `log_model=True` that uploads only at the end is for that exact purpose
> * depending on the answer to previous function, we can rethink allowed values for `log_model` (right now `True` and `all`)
> * should the "artifact removal" be on by default? I don't have a strong opinion about it.
> * how to integrate it best? I like `keep_artifacts_history` (suggested by @ohayonguy ) as an arg in the callback and a separate function of the same name that will be called when the flag is on.
> * I suggest to keep the functionality `best_1`, `best_2`… when wanting to keep top k models in a separate PR as it is more complex and opens many more questions

Regarding the network bandwidth, I agree that this might be a problem to some people.
However, logging the checkpoints only at the end of training is not that safe. In addition, if someone really has a network bandwidth issue, they can reference artifacts to their local file system instead of uploading them to wandb (the checkpoints are saved by `ModelCheckpoint` anyways.
Despite all that, I think that we should still keep the functionality to upload only at the end of training (why not?).

How about this:

- `log_model` could be either `True` or `False`. If it is set to `True`, then you can play with the way artifacts are uploaded with a different flag:
- `artifacts_history` could be `all` or `most_recent` and it is self explanatory. This flag is relevant not only for `log_model`, but for all possible artifacts. This logger might be expanded in the future to collect more kinds of artifacts, and this flag will determine whether history is preserved or not, for all artifacts.

What do you think? @borisdayma @tchaton As in all engineering, there will be a tradeoff eventually...

**Comment by tchaton** [comment]
> Ok so I really like the purpose of this PR.
> 
> I think there are a few things to decide:
> 
> * should we be concerned about bandwidth, max upload quota for users? I personally don't have that issue but the reason we have a mode `log_model=True` that uploads only at the end is for that exact purpose
> * depending on the answer to previous function, we can rethink allowed values for `log_model` (right now `True` and `all`)
> * should the "artifact removal" be on by default? I don't have a strong opinion about it.
> * how to integrate it best? I like `keep_artifacts_history` (suggested by @ohayonguy ) as an arg in the callback and a separate function of the same name that will be called when the flag is on.
> * I suggest to keep the functionality `best_1`, `best_2`… when wanting to keep top k models in a separate PR as it is more complex and opens many more questions

Hey @borisdayma, quick question about Wandb internals. Are the model uploaded within a separated thread / process ? How would be the impact of training to log all models ?

Best,
T.C

**Comment by borisdayma** [comment]
> Hey @borisdayma, quick question about Wandb internals. Are the model uploaded within a separated thread / process ? How would be the impact of training to log all models ?

It is performed in a separate process so should not impact training at all (even if the wandb process were to fail).

> How about this:
> 
> * `log_model` could be either `True` or `False`. If it is set to `True`, then you can play with the way artifacts are uploaded with a different flag:
> * `artifacts_history` could be `all` or `most_recent` and it is self explanatory. This flag is relevant not only for `log_model`, but for all possible artifacts. This logger might be expanded in the future to collect more kinds of artifacts, and this flag will determine whether history is preserved or not, for all artifacts.

That's a great idea! In addition I suggest:
* `log_model=True` logs during training by default (more convenient), so it's the same as the current `log_model='all'`
* `log_model='all'` still works but will be deprecated in the future (warning for now)
* logging only at the end needs a new flag or a new argument

**Comment by tchaton** [comment]
Hey @borisdayma, @ohayonguy,

Yes, I like the proposal too. Should we adapt this PR directly with the suggested changes ?

Best,
T.C

**Comment by borisdayma** [comment]
As you guys want! I can work on it later this week unless @ohayonguy wants to update this PR?

**Comment by ohayonguy** [comment]
> As you guys want! I can work on it later this week unless @ohayonguy wants to update this PR?

Hey guys @borisdayma @tchaton 
I can definitely work on it (and would be happy to). However it could take a while as I am currently super busy with my internship.
Sorry for the late response.

**Comment by sh3rlock14** [comment]
Hi everyone 👋🏻!
I see that this PR is now closed and the proposed feature couldn't make it into newer versions of PL: did you chose to keep the behaviour of WandbLogger as it is, or is there something new that is been developing?

Thanks for your help!

**Comment by Buckler89** [comment]
Any updates on this? Will this feature be merged?
I think this is important since logging the model during training helps protect against crashes and allows the experiment to be resumed from its last saved checkpoint.

**Comment by taczin** [comment]
+1 for this feature. Anyone wanting to train a large model will be interested in this. Due to the large model size keeping a full history of checkpoints will take up too much disk space but waiting for end of training to upload top k model checkpoints is too vulnerable to crashes.

**Comment by marvinsxtr** [comment]
In case you are looking for a simple solution, writing a custom `WandbLogger` does the trick:

```python
from lightning.pytorch.callbacks.model_checkpoint import ModelCheckpoint
from lightning.pytorch.loggers import WandbLogger

import wandb


class CustomWandbLogger(WandbLogger):
    """Removes artifacts that are not tagged as best or latest."""

    def _scan_and_log_checkpoints(self, checkpoint_callback: ModelCheckpoint) -> None:
        super()._scan_and_log_checkpoints(checkpoint_callback)

        try:
            api = wandb.Api(overrides={"project": self.experiment.project})
    
            for version in api.artifacts("model", f"model-{self.experiment.id}"):
                # Clean up all versions that don't have an alias such as 'latest'.
                if len(version.aliases) == 0:
                    version.delete()
        except HTTPError:
            print("Could not delete old WandB artifacts. Skipping...")
```
